File: app/src/com/ternaryop/photoshelf/fragment/ScheduledListFragment.java
package com.ternaryop.photoshelf.fragment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;

import com.ternaryop.photoshelf.R;
import com.ternaryop.photoshelf.adapter.PhotoShelfPost;
import com.ternaryop.tumblr.Tumblr;
import com.ternaryop.tumblr.TumblrPhotoPost;
import com.ternaryop.tumblr.TumblrPost;
import com.ternaryop.utils.AbsProgressBarAsyncTask;

public class ScheduledListFragment extends AbsPostsListFragment {
    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        
        photoAdapter.setOnPhotoBrowseClick(this);
        if (getBlogName() != null) {
            offset = 0;
            hasMorePosts = true;
            readPhotoPosts();
        }
    }

    @Override
    protected int getActionModeMenuId() {
        return R.menu.scheduled_context;
    }
    
    @Override
    protected void readPhotoPosts() {
        if (isScrolling) {
            return;
        }
        refreshUI();
        isScrolling = true;

        new AbsProgressBarAsyncTask<Void, String, List<PhotoShelfPost> >(getActivity(), getString(R.string.reading_scheduled_posts)) {
            @Override
            protected void onProgressUpdate(String... values) {
                getProgressDialog().setMessage(values[0]);
            }
            
            @Override
            protected void onPostExecute(List<PhotoShelfPost> posts) {
                super.onPostExecute(posts);
                
                if (getError() == null) {
                    photoAdapter.addAll(posts);
                    refreshUI();
                }
                isScrolling = false;
            }

            @Override
            protected List<PhotoShelfPost> doInBackground(Void... voidParams) {
                try {
                    HashMap<String, String> params = new HashMap<String, String>();
                    params.put("offset", String.valueOf(offset));
                    List<TumblrPost> photoPosts = Tumblr.getSharedTumblr(getContext())
                            .getQueue(getBlogName(), params);

                    List<PhotoShelfPost> photoList = new ArrayList<PhotoShelfPost>(); 
                    for (TumblrPost post : photoPosts) {
                        if (post.getType().equals("photo")) {
                            photoList.add(new PhotoShelfPost((TumblrPhotoPost)post,
                                    post.getScheduledPublishTime() * 1000));
                        }
                    }
                    if (photoPosts.size() > 0) {
                        // tumblr doesn't return the total posts count for scheduled posts
                        // so we set to the photoList count
                        totalPosts = photoList.size();
                        hasMorePosts = true;
                    } else {
                        totalPosts = photoAdapter.getCount() + photoList.size();
                        hasMorePosts = false;
                    }
                    return photoList;
                } catch (Exception e) {
                    e.printStackTrace();
                    setError(e);
                }
                return Collections.emptyList();
            }
        }.execute();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.scheduler, menu);

        super.onCreateOptionsMenu(menu, inflater);
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/fragment/ScheduledListFragment.kt
package com.ternaryop.photoshelf.fragment

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.ternaryop.photoshelf.R
import com.ternaryop.photoshelf.adapter.PhotoShelfPost
import com.ternaryop.photoshelf.event.CounterEvent
import com.ternaryop.photoshelf.lifecycle.Status
import com.ternaryop.photoshelf.util.post.PageFetcher

open class ScheduledListFragment : AbsPagingPostsListFragment() {
    private lateinit var viewModel: ScheduledListViewModel

    override val actionModeMenuId: Int
        get() = R.menu.scheduled_context

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        photoAdapter.counterType = CounterEvent.SCHEDULE

        viewModel = ViewModelProviders.of(this).get(ScheduledListViewModel::class.java)

        viewModel.result.observe(viewLifecycleOwner, Observer { result ->
            when (result) {
                is ScheduledListResult.Scheduled -> onFetchPosts(result)
            }
        })
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        photoAdapter.setOnPhotoBrowseClick(this)

        if (blogName != null) {
            fetchPosts(true)
        }
    }

    override val pageFetcher: PageFetcher<PhotoShelfPost>
        get() = viewModel.pageFetcher

    override fun fetchPosts(fetchCache: Boolean) {
        refreshUI()
        photoShelfSwipe.setRefreshingAndWaitingResult(true)

        val params = mapOf(
            "offset" to pageFetcher.pagingInfo.offset.toString())

        // we assume all returned items are photos, (we handle only photos)
        // if other posts type are returned, the getQueue() list size may be greater than photo list size
        viewModel.scheduled(blogName!!, params, fetchCache)
    }

    private fun onFetchPosts(result: ScheduledListResult.Scheduled) {
        when (result.command.status) {
            Status.SUCCESS -> {
                result.command.data?.also { fetched ->
                    photoShelfSwipe.setRefreshingAndWaitingResult(false)
                    photoAdapter.setPosts(fetched.list)
                    refreshUI()
                }
            }
            Status.ERROR -> {
                photoShelfSwipe.setRefreshingAndWaitingResult(false)
                result.command.error?.also { showSnackbar(makeSnake(recyclerView, it)) }
            }
            Status.PROGRESS -> {
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.scheduler, menu)

        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_refresh -> {
                clearThenReloadPosts()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
